### PR TITLE
In-App Feedback Prompt UI

### PR DESCRIPTION
**Pull request by salimbraksa**
Part of #22341

## Related PR
This PR depends on:
- https://github.com/wordpress-mobile/WordPress-iOS/pull/22050

## Description

This PR introduces the In-App Feedback Prompt UI. The prompt becomes visible after a combination of any of these actions is performed at least 20 times:
1. User taps on an unread notification
2. User publishes a post
3. User visits stats
4. User reads a post

| Positive Feedback Flow | Negative Feedback Flow |
| ------------------------ | ------------------------ |
| <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/9609223/dd365b98-152d-41c4-b847-4d52bfba5344" ></video> | <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/9609223/5e53df1d-5d27-459e-bd14-0380287c6284" ></video> |

#### Positive Feedback Flow
1. Open `WordPressAppDelegate.swift` and on line 411, set `utility.systemWideSignificantEventCountRequiredForPrompt` to 5.
2. Run Jetpack and Login.
3. Perform a combination of these actions at least 5 times:
   - User taps on an unread notification
   - User publishes a post
   - User visits stats
   - User reads a post.
4. Navigate to My Site Dashboard screen.
5. **Expect** a "Your feecback matters" alert to appear.
6. **Expect** the event `app_reviews_saw_prompt` to be tracked.
7. Tap "It's great".
8. **Expect** the event `app_reviews_liked_app` to be tracked.
9. **Expect** "Enjoying Jetpack" alert to appear with a component to rate the app.
    - _You may notice the Submit button is disabled after rating the app, the reason is that the alert is a native component and the button is disabled in dev environment._
10. Navigate to any other screen then navigate back to **My Site Dashboard**.
11. **Expect** the In-App Feedback Prompt **not** to appear.

#### Negative Feedback Flow
1. For the In-App Feedback to appear again, you can either delete, and reinstall the Jetpack to clear the cache, or force the prompt to appear programmatically:
   - Open `InAppFeedbackPromptCoordinator.swift` and on line 25, force `shouldShowPromptForAppReview` to return `true`.
2. Navigate to My Site Dashboard screen.
3. Run Jetpack and Login.
4. **Expect** a "Your feedback matters" alert to appear.
5. Tap "Not really".
6. **Expect** the event `app_reviews_didnt_like_app` to be tracked.
7. **Expect** another alert with title "Help us improve" to appear.
8. Tap "Sure".
9. **Expect**  Submit Feedback screen to appear.
10. **Expect** the event `app_reviews_feedback_screen_opened` to be tracked.
11. Write a feedback.
12. Tap "Submit".
13. **Expect** a success message to appear.
14. **Expect** the event `app_reviews_feedback_sent <feedback: The app should be improved!>` to be tracked.

#### Cancel Negative Feedback Flow
1. Trigger the In-App Feedback Prompt
2. Tap "Not Really"
3. Tap "Not Now"
4. **Expect** the event `app_reviews_declined_to_rate_app` to be tracked
5. Repeat 1 to 3
6. Tap "Sure"
7. Tap "Cancel"
8. Expect the event `app_reviews_feedback_screen_canceled` to be tracked

#### Regression: Compliance Popover
1. Make the following code changes to force the compliance popover to appear:
   - Go to `CompliancePopoverCoordinator.swift`
   - On line 32, comment this code: `defaults.didShowCompliancePopup`
   - On line 60, force the method to return `true`.
2. Run Jetpack and Login.
3. Expect the Compliance Popover to appear. 

## Regression Notes
1. Potential unintended areas of impact
The presentation of the Compliance Popover should be tested.

4. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual Testing.

5. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [x] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by mokagio** [comment]
I've been going trough PRs to update them via `git merge trunk` to include https://github.com/wordpress-mobile/WordPress-iOS/pull/22270 and run on the new Apple Silicon CI with Xcode 15.1.

I did not work on this one because it's based off another PR.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift [hunk 1]
```diff
@@ -254,11 +254,6 @@ class NotificationsViewController: UIViewController, UIViewControllerRestoration
 //            self.showInlinePrompt()
 //        }
 
-        if !unreadNotificationIds.isEmpty {
-            // Track as significant event for App Rating calculations
-            AppRatingUtility.shared.incrementSignificantEvent()
-        }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:260]: This code was supposed to increment the significant events count when the user visits the Notifications screen and has unread notifications. But that didn't work, the `unreadNotificationIds` **was empty when I visit the screen even though there are unread notifications.** 

After some digging, I realized the `unreadNotificationIds` is populated when the user switches to `Unread` tab, then it is emptied when switching to another tab. 

This behavior is buggy in my opinion, and I replaced it with the "User taps on an unread notification" action.

cc @hassaanelgarem

File: WordPress/Classes/Utility/In-App Feedback/AppRatingsUtility.swift [hunk 2]
```diff
@@ -11,7 +11,7 @@ class AppRatingUtility: NSObject {
     /// section so as to trigger the review prompt for a fairly active user who
     /// uses the app in a broad fashion.
     ///
-    @objc var systemWideSignificantEventCountRequiredForPrompt: Int = 1
+    @objc var systemWideSignificantEventCountRequiredForPrompt: Int = 5
```
Comment (salimbraksa) [inline WordPress/Classes/Utility/In-App Feedback/AppRatingsUtility.swift]: This is strange. I pushed a commit to revert this change, so locally the value of this variable should be `1`, and yet it shows `5` in the PR.

CleanShot 2024-01-10 at 14 56 58@2x

But I moved this file to a different folder, which is probably why Git got confused.

File: WordPress/Classes/ViewRelated/EEUUSCompliance/CompliancePopoverCoordinator.swift [hunk 3]
```diff
@@ -25,16 +25,20 @@ final class CompliancePopoverCoordinator: CompliancePopoverCoordinatorProtocol {
         self.defaults = defaults
     }
 
-    func presentIfNeeded() {
-        guard FeatureFlag.compliancePopover.enabled/*, !defaults.didShowCompliancePopup */else {
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/EEUUSCompliance/CompliancePopoverCoordinator.swift:29]: The change `/*, !defaults.didShowCompliancePopup */` was accidentally merged in `release/24.0`. This PR reverts the change but I'll follow up with another PR targeting 24.0 and adding a Unit Test for the `presentIfNeeded` method.

File: WordPress/Classes/Utility/In-App Feedback/AppRatingsUtility.swift (hunk 2 above)
Comment (alpavanoglu) [reply WordPress/Classes/Utility/In-App Feedback/AppRatingsUtility.swift]: I'll checkout to your branch and verify

File: WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+protocol InAppFeedbackPromptPresenting {
+    func presentIfNeeded(in controller: UIViewController) -> Bool
+}
+
+struct InAppFeedbackPromptCoordinator: InAppFeedbackPromptPresenting {
+
+    // MARK: - Properties
+
+    private let appRatingUtility: AppRatingUtility
+    private let userDefaults: UserDefaults
```
Comment (alpavanoglu) [inline WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: `UserPersistentRepository` can be used instead of `UserDefaults` as type. It simply adds an abstraction over `UserDefaults` to allow us injecting.

File: WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    private func feedbackAlert(in controller: UIViewController) -> UIAlertController {
+        let alert = UIAlertController(
+            title: Strings.FeedbackAlert.title,
+            message: Strings.FeedbackAlert.message,
+            preferredStyle: .alert
+        )
+        let yes = UIAlertAction(title: Strings.FeedbackAlert.yes, style: .default) { _ in
+            self.handlePositiveFeedback(in: controller)
+        }
```
Comment (alpavanoglu) [inline WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: I suppose there's no retain cycle as `UIAlertController` will be deallocated once the function ends. Though I thought it could make it less prone for a cycle in the future. Up to you.

Comment (salimbraksa) [reply WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: Good catch, I'll add `[weak self]` to this closure and all other UIAlertAction closures.

File: WordPress/Classes/Utility/In-App Feedback/SubmitFeedbackViewController.swift [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    private var cancellables = Set<AnyCancellable>()
+
+    // MARK: - View Lifecycle
+
+    deinit {
+        guard !feedbackWasSubmitted else {
+            return
+        }
+        WPAnalytics.track(.appReviewsCanceledFeedbackScreen)
+    }
```
Comment (alpavanoglu) [inline WordPress/Classes/Utility/In-App Feedback/SubmitFeedbackViewController.swift]: Shouldn't we do this in `viewDidDisappear` or some other view lifecycle function rather than deinit which is about memory deallocation?

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+class MySiteOverlaysCoordinator {
+
+    // MARK: - Dependencies
+
+    private var complianceCoordinator: CompliancePopoverCoordinator?
+    private var inAppFeedbackCoordinator: InAppFeedbackPromptPresenting?
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: I am not sure what kind of structure you have followed but if VC retains Coordinators (whether directly or via ViewModel), these should be `unowned` variables rather than strongly held. Since we would wish the children Coordinators to be deallocated once their respective VCs are.

I just wanted to point that out.

**Comment by alpavanoglu** [comment]
## 🧪 Test Results 🧪 

### Positive Feedback Flow
❓ Since I cannot Submit (on simulator), I had to press Cancel. After that, the view is dismissed. And the rest of steps passed.

### Negative Feedback Flow
✅ All steps work as expected.

### Cancel Negative Feedback Flow
✅ All steps work as expected.

### Regression: Compliance Popover
✅ All steps work as expected.

File: WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift (hunk 4 above)
Comment (salimbraksa) [reply WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: Resolved in cb10a4ac12fe6ec8aba3cdfbd1baa19f3d48e994. I've removed the `userDefaults` property as it was not used.

File: WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift (hunk 5 above)
Comment (salimbraksa) [reply WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: I got this error when adding `[weak self]`:
```
'weak' may only be applied to class and class-bound protocol types, not 'InAppFeedbackPromptCoordinator'
```
That makes sense since `InAppFeedbackPromptCoordinator` is a `struct`.

File: WordPress/Classes/Utility/In-App Feedback/SubmitFeedbackViewController.swift (hunk 6 above)
Comment (salimbraksa) [reply WordPress/Classes/Utility/In-App Feedback/SubmitFeedbackViewController.swift]: Resolved in a9dd8ab5d8a26f8e477d71424137e8cd967b854d

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift (hunk 7 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: > these should be unowned variables rather than strongly held. Since we would wish the children Coordinators to be deallocated once their respective VCs are.

Their respective VCs don't hold a reference to these children coordinators, so if we mark them as `unowned` , they will be deallocated immediately.

I have considered instantiating those children coordinators in `MySiteOverlaysCoordinator.presentOverlayIfNeeded`, but I'm planning to write unit tests for `MySiteOverlaysCoordinator` and mock those children coordinators dependencies.

File: WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift (hunk 5 above)
Comment (alpavanoglu) [reply WordPress/Classes/Utility/In-App Feedback/InAppFeedbackPromptCoordinator.swift]: Oh, that actually may not be such a great idea as it contains other reference types. So when you pass it around, its value type properties will be copied but the reference types will be modified throughout every copy.

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift (hunk 7 above)
Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: I see. So how are the Coordinators retained? A parent coordinator retains a child coordinator and that one retains its children coordinators? If that's the case, do we also need to deallocate them once the screen is dismissed?

So if Coordinator A (CA) is the parent:

CA -> CB -> CC and the ViewController that is linked to CC is dismissed, is there a mechanism to deallocate CC in place?

**Review by alpavanoglu — APPROVED** [review]
Approving to not block the progress. However I think that my last 2 comments might be important. If after considering them, you think it is safe to go ahead, please feel free to merge.

Thanks for taking care of my earlier comments!

File: WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift (hunk 7 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: @alpavanoglu `MySiteViewController` retains `MySiteOverlaysCoordinator`, which in turns retains those child coordinators.

MySiteVC > MySiteOverlaysCoordinator > Child Coordinators

Also, none of the coordinators retain a View Controller. And the coordinators will be deallocated when MySiteVC is deallocated.

Keep in mind that the coordinators are only responsible for displaying their respective VCs, but they don't retain them.

Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: Okay. So if a child coordinator's ViewController is dismissed, the child coordinator itself will not be deallocated because it is being referenced by the parent coordinator. If you think in terms of navigation stack, this would be like a whole flow being retained until their root VC is dismissed/deallocated.

Because of this problem, I personally prefer that each respective VC retains their Coordinator rather than a parent Coordinator retaining children coordinators. Otherwise coordinators have to replicate a deallocation chain that is provided by Navigation Stack out of the box.

Of course all of this is with the assumption that Coordinators are `class` and not `struct`. Since they often contain states that change over its lifetime, I find class to be a better fit.

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: > I personally prefer that each respective VC retains their Coordinator rather than a parent Coordinator retaining children coordinators.

I think this is not always possible. In cases like the In-App Feedback coordinator, where the VC is just a `UIAlertViewController`, we have to create a subclass to retain the Coordinator.

Additionally, the `InAppFeedbackCoordinator` was inspired by the `CompliancePopoverCoordinator`, and the latter doesn't retain its VC. 

I'm not against your suggestion, but I still don't know what to change exactly. Could you provide a code example to clarify this approach?

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: > If you think in terms of navigation stack, this would be like a whole flow being retained until their root VC is dismissed/deallocated.

I understand, but this is not happening in our case. `MySiteViewController` retains **ONLY** the coordinators of the screens it is responsible for presenting.

The coordinators are just an alternative to writing the presentation logic within the `MySiteViewController`.

Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Blog/My Site/MySiteOverlaysCoordinator.swift:8]: I see. The structure confused me to think otherwise. `MySiteOverlaysCoordinator` is merely a helper to ensure only one coordinator is active at a given time. **No action needed then.**

> Additionally, the InAppFeedbackCoordinator was inspired by the CompliancePopoverCoordinator, and the latter doesn't retain its VC.

Yeah I think it is better that the Coordinator does not retain VCs. Maybe I wasn't articulate enough. Navigation stack already retains VCs down the chain. When Coordinators do it too, we need to deal with deallocating them.
